libp2p/src/builder.rs

```diff
@@ -0,0 +1,282 @@
+use libp2p_core::{muxing::StreamMuxerBox, Transport};
+use std::marker::PhantomData;
+
+pub struct TransportBuilder {
+    keypair: libp2p_identity::Keypair,
+}
+
+impl TransportBuilder {
+    pub fn new(keypair: libp2p_identity::Keypair) -> Self {
+        Self { keypair }
+    }
+
+    #[cfg(feature = "async-std")]
+    pub fn with_async_std(self) -> TcpBuilder<AsyncStd> {
+        TcpBuilder {
+            keypair: self.keypair,
+            phantom: PhantomData,
+        }
+    }
+
+    #[cfg(feature = "tokio")]
+    pub fn with_tokio(self) -> TcpBuilder<Tokio> {
+        TcpBuilder {
+            keypair: self.keypair,
+            phantom: PhantomData,
+        }
+    }
+}
+
+pub struct TcpBuilder<P> {
+    keypair: libp2p_identity::Keypair,
+    phantom: PhantomData<P>,
+}
+
+#[cfg(all(feature = "async-std", feature = "tcp"))]
+impl TcpBuilder<AsyncStd> {
+    pub fn with_tcp(self) -> RelayBuilder<AsyncStd, impl BuildableTransport> {
+        RelayBuilder {
+            transport: libp2p_tcp::async_io::Transport::new(
+                libp2p_tcp::Config::new().nodelay(true),
+            )
+            .upgrade(libp2p_core::upgrade::Version::V1)
```

Comment:
I think we should default to V1Lazy here in preparation of always making this the default.